core/src/main/java/org/moditect/commands/AddModuleInfo.java

```diff
@@ -99,62 +101,65 @@ public void run() {
-
-                Path manifestPath = zipfs.getPath("META-INF/MANIFEST.MF");
-                Manifest manifest;
-                if (Files.exists(manifestPath)) {
-                    manifest = new Manifest(Files.newInputStream(manifestPath));
+        long lastModifiedTime = toFileTime(timestamp).toMillis();
+        // brute force copy all entries
+        try (JarFile jarFile = new JarFile(inputJar.toAbsolutePath().toFile());
+                JarOutputStream jarout = new JarOutputStream(Files.newOutputStream(outputJar.toAbsolutePath(), TRUNCATE_EXISTING))) {
+            Enumeration<JarEntry> entries = jarFile.entries();
+            while (entries.hasMoreElements()) {
+                JarEntry inputEntry = entries.nextElement();
+
+                // manifest requires extra care due to MRJARs
+                if (MANIFEST_ENTRY_NAME.equals(inputEntry.getName()) && jvmVersion != null) {
+                    Manifest manifest = jarFile.getManifest();
+                    if (null == manifest) {
+                        manifest = new Manifest();
+                        manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+                    }
+                    manifest.getMainAttributes().put(new Attributes.Name("Multi-Release"), "true");
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    manifest.write(baos);
+
+                    JarEntry outputEntry = new JarEntry(inputEntry.getName());
+                    outputEntry.setTime(lastModifiedTime);
+                    jarout.putNextEntry(outputEntry);
+                    jarout.write(baos.toByteArray(), 0, baos.size());
+                    jarout.closeEntry();
                 }
                 else {
-                    manifest = new Manifest();
-                    manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
+                    // copy entry as is, set tiemstamp
+                    JarEntry outputEntry = new JarEntry(inputEntry.getName());
+                    outputEntry.setTime(lastModifiedTime);
+                    jarout.putNextEntry(outputEntry);
+                    InputStream in = jarFile.getInputStream(inputEntry);
+                    copy(in, jarout);
+                    jarout.closeEntry();
                 }
+            }
 
-                manifest.getMainAttributes().put(new Attributes.Name("Multi-Release"), "true");
-                try (OutputStream manifestOs = Files.newOutputStream(manifestPath, StandardOpenOption.TRUNCATE_EXISTING)) {
-                    manifest.write(manifestOs);
-                }
-                setTimes(manifestPath, lastModifiedTime);
+            // copy module descriptor
+            JarEntry outputEntry = new JarEntry("module-info.class");
+            if (jvmVersion != null) {
+                outputEntry = new JarEntry("META-INF/versions/" + jvmVersion + "/module-info.class");
```

Comment:
Unconditionally appending `module-info.class` at the end of the JAR can lead to problems if the original JAR already contains a `module-info.class`.

This can happen in practice if the user runs the build twice and `AddModuleInfo` is executed twice in a row.

Reply:
Indeed, good catch. Removing any existing entries for the descriptor should also solve https://github.com/moditect/moditect/issues/215

Reply:
done